rest-api/offline/sync_consent_files.py

```diff
@@ -0,0 +1,151 @@
+  for participant_data in _iter_participants_data():
+    kwargs = {
+      "source_bucket": SOURCE_BUCKET,
+      "destination_bucket": org_data_map[participant_data.org_id].bucket_name,
+      "participant_id": participant_data.participant_id,
+      "google_group": participant_data.google_group or DEFAULT_GOOGLE_GROUP
+    }
+    deferred.defer(
+      cloudstorage_copy_objects,
+      '/{source_bucket}/Participant/P{participant_id}/'.format(**kwargs),
+      '/{destination_bucket}/Participant/{google_group}/P{participant_id}/'.format(**kwargs)
+    )
+
+
+class GoogleSheetCSVReader(csv.DictReader):
+
+  def __init__(self, sheet_id, gid=0, *args, **kwds):
+    self._sheet_id = sheet_id
+    self._gid = gid
+    response = requests.get(self._get_sheet_url(self._sheet_id, self._gid))
+    csv.DictReader.__init__(self, StringIO.StringIO(response.text), *args, **kwds)
+
+  @staticmethod
+  def _get_sheet_url(sheet_id, gid):
+    return (
+      "https://docs.google.com/spreadsheets/d/{id}/export"
+      "?format=csv"
+      "&id={id}"
+      "&gid={gid}"
+    ).format(
+      id=sheet_id,
+      gid=gid
+    )
+
+
+def _load_org_data_map(sheet_id):
+  # parse initial mapping
+  org_data_map = {
+    row.get(COLUMN_ORG_ID): OrgData(
+      row.get(COLUMN_ORG_ID),
+      row.get(COLUMN_AGGREGATING_ORG_ID),
+      row.get(COLUMN_BUCKET_NAME)
+    )
+    for row in GoogleSheetCSVReader(sheet_id)
+    if row.get(COLUMN_ORG_STATUS) == ORG_STATUS_ACTIVE
+  }
+  # apply transformations that require full map
+  return {
+    org_data.org_id: _org_data_with_bucket_inheritance_from_org_data(org_data_map, org_data)
+    for org_data in org_data_map.values()
+  }
+
+
+def _org_data_with_bucket_inheritance_from_org_data(org_data_map, org_data):
+  return OrgData(
+    org_data.org_id,
+    org_data.aggregate_id,
+    (
+      org_data.bucket_name
+      if org_data.bucket_name
+      else org_data_map[org_data.aggregate_id].bucket_name
+    )
+  )
+
+
+PARTICIPANT_DATA_SQL = sqlalchemy.text("""
+select
+  participant.participant_id,
+  site.google_group,
+  organization.external_id
+from participant
+left join organization
+  on participant.organization_id = organization.organization_id
+left join site
+  on participant.site_id = site.site_id
+left join participant_summary summary
+  on participant.participant_id = summary.participant_id
+where summary.consent_for_electronic_health_records = 1
+  and summary.consent_for_study_enrollment = 1
+""")
+
+
+def _iter_participants_data():
+  with database_factory.make_server_cursor_database().session() as session:
+    for row in session.execute(PARTICIPANT_DATA_SQL):
+      yield ParticipantData(*row)
+
+
+def cloudstorage_copy_objects(source, destination):
+  """Copies all objects matching the source to the destination
+
+  Both source and destination use the following format: /bucket/prefix/
+  """
+  for file_stat in cloudstorage.listbucket(source):
+    cloudstorage.copy2(
```

Comment:
My only concern is we’ll always copy all files. At the moment it’s about 800,000 I believe.

Reply:
The newer commit now gets and checks the destination file's etag (md5 hash) to determine if it should copy.
This will save some bandwidth for already existing files but this still has 1 extra api call to get the `GCSFileStat` for each potential destination bucket path.